### PR TITLE
For upstream

### PR DESCRIPTION
**Pull request by LiuZhiguang001**
Patches to parse HOB about PCI, Acpi and Smbios

### CODE REVIEW

File: UefiPayloadPkg/Include/Guid/PldPciRootBridgeInfoHob.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef __PLD_PCI_ROOT_BRIDGE_INFO_HOB_H__
+#define __PLD_PCI_ROOT_BRIDGE_INFO_HOB_H__
+
+#include <IndustryStandard/Pci.h>
+#include <Library/PciHostBridgeLib.h>
+
```
Comment (niruiyu) [inline UefiPayloadPkg/Include/Guid/PldPciRootBridgeInfoHob.h:14]: Do we need a #pragma pack(1)?

File: MdeModulePkg/Include/Guid/SmBiosTableHob.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+Copyright (c) 2021, Intel Corporation. All rights reserved.<BR>
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef _SMBIOS_TABLE_HOB_H_
+#define _SMBIOS_TABLE_HOB_H_
+
+#include <Uefi.h>
```
Comment (niruiyu) [inline MdeModulePkg/Include/Guid/SmBiosTableHob.h:12]: Is it ok to just include <Base.h>?
We don't want to let PLD interface depend on UEFI types.

File: MdeModulePkg/Include/Guid/SmBiosTableHob.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+typedef struct {
+  EFI_PHYSICAL_ADDRESS          SmBiosTableAddress;
```
Comment (niruiyu) [inline MdeModulePkg/Include/Guid/SmBiosTableHob.h]: I understand Laszlo proposed to use EFI_PHYSICAL_ADDRESS other than UINT64. But using the EFI_XXX type requires dependency on UEFI types.

@mauricema , @gdong1 , @vincent-j-zimmer , what's your thought?

File: MdeModulePkg/Include/Guid/SmBiosTableHob.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+} PLD_SMBIOS_TABLE_HOB;
+
+extern EFI_GUID gPldSmbios3TableGuid;
```
Comment (niruiyu) [inline MdeModulePkg/Include/Guid/SmBiosTableHob.h]: Change EFI_GUID to GUID type.

File: MdeModulePkg/Include/Guid/SmBiosTableHob.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+extern EFI_GUID gPldSmbiosTableGuid;
+#endif
```
Comment (niruiyu) [inline MdeModulePkg/Include/Guid/SmBiosTableHob.h]: Please add end of line

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c [hunk 6]
```diff
@@ -1408,6 +1408,242 @@ SmbiosTableConstruction (
...
+
+  @param  SmbiosTable   The SMBIOS_TABLE_ENTRY_POINT to validate.
+
+  @retval TRUE           SMBIOS table entry point is valid.
+  @retval FALSE          SMBIOS table entry point is malformed.
+
+**/
+STATIC
+BOOLEAN
+ValidateSmbios20Table(
```
Comment (niruiyu) [inline MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: Change name to "IsValidSmbios20Table"

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c [hunk 7]
```diff
@@ -1408,6 +1408,242 @@ SmbiosTableConstruction (
...
+  IN SMBIOS_TABLE_ENTRY_POINT      *SmbiosTable
+) {
+  UINT8 Checksum;
+
+  if (CompareMem (SmbiosTable->AnchorString, "_SM_", 4) != 0) {
+    return FALSE;
```
Comment (niruiyu) [inline MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c:1429]: Any existing definition for "\_SM\_"? We should not hard code here.

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c [hunk 8]
```diff
@@ -1408,6 +1408,242 @@ SmbiosTableConstruction (
...
+  }
+  if (SmbiosTable->EntryPointLength < 0x1E) {
```
Comment (niruiyu) [inline MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: "0x1E" is also hardcode value. please avoid that.

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c [hunk 9]
```diff
@@ -1408,6 +1408,242 @@ SmbiosTableConstruction (
...
+  }
+
+  Checksum = CalculateSum8((UINT8 *) SmbiosTable,
+    SmbiosTable->EntryPointLength);
+  if (Checksum != 0) {
+    return FALSE;
+  }
+
+  Checksum = CalculateSum8((UINT8 *) SmbiosTable + 0x10,
+    SmbiosTable->EntryPointLength - 0x10);
```
Comment (niruiyu) [inline MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: What does this checksum calculation mean? If there is a reason, please add comments.

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c [hunk 10]
```diff
@@ -1408,6 +1408,242 @@ SmbiosTableConstruction (
...
+  @retval FALSE          SMBIOS table entry point is malformed.
+
+**/
+STATIC
+BOOLEAN
+ValidateSmbios20Table(
+  IN SMBIOS_TABLE_ENTRY_POINT      *SmbiosTable
+) {
+  UINT8 Checksum;
+
```
Comment (niruiyu) [inline MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c:1427]: Put comments for every if check to tell what you are checking.

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c [hunk 11]
```diff
@@ -1408,6 +1408,242 @@ SmbiosTableConstruction (
...
+  SMBIOS_STRUCTURE_POINTER      Smbios;
+  EFI_HOB_GUID_TYPE             *GuidHob;
+  PLD_SMBIOS_TABLE_HOB          *SmBiosTableAdress;
+
+  //
+  // Scan for existing SMBIOS tables from gPldSmbios3TableGuid Guid Hob
+  //
+  GuidHob = GetFirstGuidHob (&gPldSmbios3TableGuid);
+  if (GuidHob != NULL) {
+    if (sizeof(PLD_SMBIOS_TABLE_HOB) + sizeof(EFI_HOB_GUID_TYPE) <= GuidHob->Header.HobLength) {
```
Comment (niruiyu) [inline MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: Use GET_GUID_HOB_DATA_SIZE

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c [hunk 12]
```diff
@@ -1408,6 +1408,242 @@ SmbiosTableConstruction (
...
+      SmBiosTableAdress = (PLD_SMBIOS_TABLE_HOB*)GET_GUID_HOB_DATA (GuidHob);
+      Smbios30Table = (SMBIOS_TABLE_3_0_ENTRY_POINT  *)SmBiosTableAdress->SmBiosTableAddress;
```
Comment (niruiyu) [inline MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: two spaces

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c [hunk 13]
```diff
@@ -1426,7 +1662,7 @@ SmbiosDriverEntryPoint (
   IN EFI_SYSTEM_TABLE     *SystemTable
   )
 {
-  EFI_STATUS            Status;
+  EFI_STATUS                    Status;
```
Comment (niruiyu) [inline MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: What's the reason of this change?

**Review by niruiyu — CHANGES_REQUESTED** [review]

**Review by niruiyu — APPROVED** [review]

File: MdeModulePkg/Include/Guid/AcpiTableHob.h [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+**/
+
+#ifndef _ACPI_TABLE_HOB_H_
+#define _ACPI_TABLE_HOB_H_
+
+#include <Uefi.h>
+
+typedef struct {
+  EFI_PHYSICAL_ADDRESS          Rsdp;
```
Comment (niruiyu) [inline MdeModulePkg/Include/Guid/AcpiTableHob.h:16]: Similar comments as that for SmBiosTableHob.h regarding EFI dependency.

File: MdeModulePkg/Include/Guid/AcpiTableHob.h [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+} PLD_ACPI_TABLE_HOB;
+
+extern EFI_GUID gPldAcpiTableGuid;
```
Comment (niruiyu) [inline MdeModulePkg/Include/Guid/AcpiTableHob.h]: Use GUID type.

File: MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c [hunk 16]
```diff
@@ -1689,6 +1699,113 @@ ChecksumCommonTables (
...
+    // XsdtAddress is zero, then we use Rsdt which has 32 bit entry
+    //
+    Rsdt = (EFI_ACPI_DESCRIPTION_HEADER*) (UINTN) Rsdp->RsdtAddress;
+    EntrySize = sizeof (UINT32);
+  }
+  Count = (Rsdt->Length - sizeof (EFI_ACPI_DESCRIPTION_HEADER))/EntrySize;
+
+  for (Index = 0; Index < Count; Index++){
+    ChildTableAddress = 0;
+    CopyMem(&ChildTableAddress, (VOID*)((UINTN) Rsdt + EntrySize * Index + sizeof (EFI_ACPI_DESCRIPTION_HEADER)), EntrySize);
```
Comment (niruiyu) [inline MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c]: CopyMem(&ChildTableAddress, (UINT8 *) (Rsdt + 1) + EntrySize * Index, EntrySize);

File: UefiPayloadPkg/UefiPayloadEntry/UefiPayloadEntry.c [hunk 17]
```diff
@@ -235,6 +235,7 @@ BuildHobFromBl (
   EFI_PEI_GRAPHICS_DEVICE_INFO_HOB GfxDeviceInfo;
   EFI_PEI_GRAPHICS_DEVICE_INFO_HOB *NewGfxDeviceInfo;
   PLD_SMBIOS_TABLE_HOB             *SmBiosTableAdress;
+  PLD_ACPI_TABLE_HOB               *AcpiTableAdress;
```
Comment (niruiyu) [inline UefiPayloadPkg/UefiPayloadEntry/UefiPayloadEntry.c]: Please AcpiTableHob as the name.
Change to SmBiosTableHob as well.

File: MdeModulePkg/Include/Guid/AcpiTableHob.h (hunk 14 above)
Comment (gdong1) [reply MdeModulePkg/Include/Guid/AcpiTableHob.h:16]: Here we use EFI_PYSICAL_ADDRESS instead of UINT64, it is not same with universal spec. so do we want to change the spec?

File: MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c [hunk 18]
```diff
@@ -552,13 +558,17 @@ AddTableToList (
...
-    Status = gBS->AllocatePages (
-                    AllocateMaxAddress,
-                    EfiACPIMemoryNVS,
-                    EFI_SIZE_TO_PAGES (CurrentTableList->TableSize),
-                    &AllocPhysAddress
-                    );
+    if (IsFromHob){
+      AllocPhysAddress = (UINTN)Table;
+    } else {
+      ASSERT ((EFI_PAGE_SIZE % 64) == 0);
```
Comment (gdong1) [inline MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c]: Remember Maurice mentioned this is not required.

File: MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c [hunk 19]
```diff
@@ -1689,6 +1699,113 @@ ChecksumCommonTables (
...
+                &TableKey
+                );
+      Status = AddTableToList (
+                AcpiTableInstance,
+                (VOID *) (UINTN)((EFI_ACPI_3_0_FIXED_ACPI_DESCRIPTION_TABLE*)ChildTable)->Dsdt,
+                TRUE,
+                Version,
+                TRUE,
+                &TableKey
+                );
```
Comment (gdong1) [inline MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c]: For all these AddTableToList (), we could remove "Status =" since we didn't check the status.
and we could put them in one line for the first one to makes this patch looks shorter.
we could also put the other 2 into 1 line by define a variable for the second parameter.

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c (hunk 9 above)
Comment (gdong1) [reply MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: yes. it would great if adding comment for  Intermediate Checksum  and EntryPoint Structure Checksum.

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c [hunk 20]
```diff
@@ -1408,6 +1408,242 @@ SmbiosTableConstruction (
...
+  if (GuidHob != NULL) {
+    if (sizeof(PLD_SMBIOS_TABLE_HOB) + sizeof(EFI_HOB_GUID_TYPE) <= GuidHob->Header.HobLength) {
+      SmBiosTableAdress = (PLD_SMBIOS_TABLE_HOB*)GET_GUID_HOB_DATA (GuidHob);
+      SmbiosTable = (SMBIOS_TABLE_ENTRY_POINT  *)SmBiosTableAdress->SmBiosTableAddress;
+      if (ValidateSmbios20Table(SmbiosTable)) {
+        Smbios.Raw = (UINT8 *)(UINTN)SmbiosTable->TableAddress;
+        Status = ParseAndAddExistingSmbiosTable(ImageHandle, Smbios, SmbiosTable->TableLength);
+        if (EFI_ERROR (Status)) {
+          DEBUG ((DEBUG_ERROR, "RetrieveSmbiosFromHob: Failed to parse preinstalled tables from gPldSmbiosTableGuid Guid Hob\n"));
+          Status = EFI_SUCCESS;
```
Comment (gdong1) [inline MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: From function description, it should return error since no SMBIOS table is really added, right?

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c [hunk 21]
```diff
@@ -1408,6 +1408,242 @@ SmbiosTableConstruction (
...
+        }
+      }
+    }
+  }
+  return EFI_SUCCESS;
```
Comment (gdong1) [inline MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: If both HOB are not found, from the function description, it should return EFI_NOT_FOUND by default?

File: MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c (hunk 18 above)
Comment (LiuZhiguang001) [reply MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c]: Yes. This is an old pull request, pull request #5 . Acutally, Maurice mentioned it in pull request #6, and I have changed it and pull request #6 is already merged in, so we can't see it in pull request list. It seems like when I update the code, I should override the old pull request instead of creat a new one.

File: MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c (hunk 19 above)
Comment (LiuZhiguang001) [reply MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c]: Yes, I agree.

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c (hunk 9 above)
Comment (LiuZhiguang001) [reply MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: It is already changed in new pull request

File: MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c (hunk 20 above)
Comment (LiuZhiguang001) [reply MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c]: yes, I agree about the return status comments, and will change it

File: MdeModulePkg/Include/Guid/AcpiTableHob.h (hunk 14 above)
Comment (LiuZhiguang001) [reply MdeModulePkg/Include/Guid/AcpiTableHob.h:16]: I remenber Ray has a concern that "EFI_PHYSICAL_ADDRESS" looks like it is UEFI related, while our spec should be more independent.

File: MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c (hunk 18 above)
Comment (gdong1) [reply MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c]: I only see 1 pull request there from for_upstream branch. It looks your #6 is not shown in the pull request. yes, better use same branch to update the patch.

**Comment by LiuZhiguang001** [comment]
I just update my branch, please help review
